rsl/rsl.go

```diff
@@ -39,37 +43,132 @@ func InitializeNamespace() error {
 	return nil
 }
 
-// AddEntry adds an RSL entry to the log for the ref and hash passed in.
-func AddEntry(refName string, commitID plumbing.Hash, sign bool) error {
+type RSLEntry struct {
+	RefName  string
+	CommitID plumbing.Hash
+	Revoke   []plumbing.Hash
```

Comment:
My current understanding of the RSL (please correct me if I'm wrong) is we're attaching a list of revoked revisions to a RSL object on the tuf ref that references a {ref, commit} state in a "main branch" (we may want to add comments for each of these fields).

To figure out a complete RSL state, do we need to walk the entire tuf ref, or is looking at the latest HEAD good enough? (i think based on the command this only appends new revocations?)

I'm worried about operations needing to walk the entire RSL tree for each operation any time we need to make a query - maybe we can precompute this and reference it as it's own blob? 🤔

Reply:
I think there are a few different questions there, let me try and answer them.

1. The RSL is stored at `refs/gittuf/reference-state-log`. It points to the tip commit, the latest RSL entry available.
2. A standard RSL entry (not a revocation) contains `RefName` and `CommitID`. `RefName` can be any ref in the repository, using the full `refs/<>/` format. The `CommitID` is the object that ref is supposed to be at at the time of signing the entry.
3. A revocation entry includes the identifiers of one or more prior RSL entries. The revoking entry and the entries revoked must all have identical `RefName`s.
4. The tip of the RSL namespace points to the latest entry. So looking at the tip isn't enough to figure out the expected states of all tracked refs, it is necessary to walk back. I'm certainly open to optimize this, I had, for example, in the prototype a local-only record of the last trusted entry for each red tracked on that copy of the repository. I think we have to threat model that though. I suspect that `GetEntry` results can at least be locally cached when performing an operation that has a lot of those calls.

Reply:
👍 cool, that tracks with what I had in my head. Thanks for confirming!

I'm still a bit concerned about needing to walk the entire graph (https://github.com/torvalds/linux would be a good repo to simulate on 🤣 ), but I'm okay with deferring on solving this for now.

My thinking was have ref to a blob that contains the entire computed state:

![image](https://user-images.githubusercontent.com/1844673/224376030-ebfd2de3-f6c9-40ef-9e59-64eb86e4183e.png)

(not sure if this gives us the same threat model properties though)

Reply:
> I'm still a bit concerned about needing to walk the entire graph (https://github.com/torvalds/linux would be a good repo to simulate on :rofl: ), but I'm okay with deferring on solving this for now.

I was quite worried about that as well but @JustinCappos rightly pointed out that those repositories are outliers. Also, I suspect clients will rarely have to walk the full history or anything close to it. Most walks will be to update local copies from what's available upstream.

> My thinking was have ref to a blob that contains the entire computed state:

Originally, I was planning to include a blob for each entry but I dismissed it as all the information for any one entry can easily be stored in the commit itself. Pre computed states for all refs would be nice but I think validating any entry present in the latest precomputed blob would still require walking back the tree to when the change was made.